Create enpoint for registring a  new subscription #27

Rolland0000 opened the pull request:
issue #27
- Create a handler for subscription
- Create /api/subscriptions enpoint for new subscription
- Store subscriber data in a HashMap (like data base)

Rolland0000 committed:
feat(dependencies): Implement integration test for health_check handler.

- Created health check handler in src/handlers/health.rs
- Added health check test in tests/health_check_test.rs

Rolland0000 committed:
feat(database): Implement user subscription and database management

- Add database module with FormUsers struct for user data
- Implement subscribe and get_subscribers handlers
- Create UserDB type using Arc<Mutex<HashMap>> for thread-safe storage
- Update CI workflow to include code coverage job
- Add Serde dependency for data serialization
- Modify lib.rs to support new subscription routes
- Add initial health check test for subscription functionality
Introduces basic database management and user subscription features with thread-safe in-memory storage mechanism.

Rolland0000 committed:
feat(database): Add database migration and initialization script for subscriptions

- Create initial migration for subscriptions table with UUID, email, name, and timestamp
- Add init_db.sh script to automate PostgreSQL database setup and migration
- Update .gitignore to exclude .cargo directory
- Implement database creation and migration using sqlx
- Provide robust error handling and dependency checks in initialization script

Rolland0000 committed:
refactor(build): Update linker configuration and dependency management

- Modify .cargo/config.toml to use clang and lld linker
- Remove mold linker configuration for x86_64-unknown-linux-gnu
- Comment out macOS-specific linker configuration
- Update Cargo.lock with new dependency versions
- Add several new dependencies to improve project configuration and functionality
- Prepare project for more robust build and linking performance

Rolland0000 committed:
refactor(tests): Reorganize health check test files

- Move health check test implementation from `health_check.rs` to `health_check_test.rs`
- Improve test file organization and reduce code duplication

Rolland0000 committed:
chore(ci): Refactor code coverage and workflow configuration

- Move code coverage script to dedicated shell script
- Update CI workflow to use new code coverage script
- Modify database documentation for improved readability
- Simplify code coverage generation and threshold checking process
- Improves CI/CD pipeline by centralizing code coverage logic

Rolland0000 committed:
refactor(dependencies): Update project dependencies and clean up unused packages

- Remove unused dependencies like `addr2line`, `adler2`, and `backtrace`
- Update core dependencies to latest versions (axum, hyper, bitflags, etc.)
- Clean up Cargo.lock to reduce unnecessary package entries
- Improve overall project dependency management and reduce bloat

Rolland0000 committed:
chore(ci): Enhance code coverage and dependency management

- Add .cargo/audit.toml for dependency vulnerability tracking
- Update CI workflow to generate and validate code coverage
- Modify .gitignore to exclude specific .cargo configuration
- Update Cargo.toml dependency version constraint for sqlx

Rolland0000 committed:
chore(security): Update Cargo audit configuration

- Remove detailed advisory ignore configuration
- Simplify advisory ignore entry for RUSTSEC-2023-0071

Rolland0000 committed:
feat(handlers): Enhance subscription and user management functionality

- Update FormUsers struct to use more standard field names (name, email)
- Modify subscribe handler to return only status code
- Improve get_subscribers handler to retrieve and return user details
- Add new test case for subscription endpoint with valid form data
- Create test_client.rs example for demonstrating API interactions
- Refactor database and handler code for improved clarity and consistency

Rolland0000 committed:
- Enhance code consistency across test and example files

Rolland0000 committed:
refactor(tests): Simplify health check test dependencies

- Remove unused import for `FormUsers` from database module

Hermann-Core commented on .cargo/config.toml:
```diff
@@ -1,6 +1,5 @@
 [target.x86_64-unknown-linux-gnu]
-linker = "clang"
-rustflags = ["-C", "link-arg=-fuse-ld=/usr/bin/mold"]
+rustflags = ["-C", "linker=clang", "-C", "link-arg=-fuse-ld=lld"]
```
`mold` is more faster that `lld`.

Rolland0000 committed:
chore(build): Update build configuration and dependencies

- Disable cargo audit advisory check for RUSTSEC-2023-0071
- Update linker configuration to use mold instead of lld for faster builds
- Enable macOS Darwin target configuration with lld linker
- Refactor CI pipeline to use cargo-llvm-cov for code coverage instead of tarpaulin
- Update coverage threshold check to 70% with improved error handling
- Remove push trigger from CI workflow, keep only pull_request events
- Enable cargo machete metadata checking in CI pipeline
- Update multiple dependencies including axum, bytes, hyper, and indexmap
- Add tower dependency to project dependencies
- Remove .cargo/config.toml from .gitignore to track linker configuration
- Add lcov.info coverage report file
- Improve health check and subscription handlers
- Update startup configuration for better build performance
